### PR TITLE
feat: create auto pickup command

### LINKED ISSUE #7: Create Pickup Command
## Auto-Pickup Command
The Auto-Pickup command should run the intake while moving towards and over the nearest ball to pick it up. Should be a standard command that runs a Parallel command group within initialize method (right click on the commands folder within vscode and click the "create new class/command" option) 

#### Constructor
Within the constructor, the drivetrain, intake, and BallCamera objects should all be passed as parameters and stored in as object variables. 

#### Initialize method
Within the initialize method store the return value of a call to the `getTrajectory` method in another object variable. If that trajectory is non-`null` (meaning the goal was detected), store and schedule a command that does the following as object variable:

- Run `FollowTrajectory` Command with Trajectory returned by the `getTrajectory` method in the GoalCamera subsystem (currently stored in an instance variable), this will move the robot over the ball. While this is happening run the intake to pick up the ball.

#### End method
Within the overridden end method cancel the command in the instance field, meaning that the command is finished.

#### Is Finished method
If the trajectory is `null` the function should return true (meaning, there is no goal detected and the command ends before it even begins). After that statement return whether or not the command stored has finished. 
_The reason for this ordering is to first verify that the command is not null to avoid a null pointer exception._

#### Helpful Docs
- [Commands](https://docs.wpilib.org/en/stable/docs/software/commandbased/commands.html)
- [Command Groups](https://docs.wpilib.org/en/stable/docs/software/commandbased/command-groups.html)
- [Command Scheduler](https://docs.wpilib.org/en/stable/docs/software/commandbased/command-scheduler.html)

### PR DESCRIPTION
**Pull request by Catalyst4222**
This PR creates an AutoPickup command, which gets a trajectory from the BallCamera and runs the intake while following that trajectory

closes #7

### CODE REVIEW

File: src/main/java/frc/robot/commands/AutoPickup.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  private Trajectory trajectory;
+  private ParallelDeadlineGroup commandGroup;
+
+  /** Creates a new AutoPickup. */
+  public AutoPickup(Drivetrain driveTrain, Intake intake, BallCamera camera) {
+    this.driveTrain = driveTrain;
+    this.intake = intake;
+    this.camera = camera;
+
+    this.trajectory = null;
```
Comment (paulelliott) [inline src/main/java/frc/robot/commands/AutoPickup.java:34]: I don't think you need to initialize these to null here. That should be the case by default.

File: src/main/java/frc/robot/commands/AutoPickup.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    this.commandGroup.schedule();
+  }
+
+
+  @Override
+  public boolean isFinished() {
+    if (this.trajectory == null || this.commandGroup == null) {
+      return true;
+    }
+    return this.commandGroup.isFinished();
```
Comment (paulelliott) [inline src/main/java/frc/robot/commands/AutoPickup.java:63]: I think you could use a short-circuit here and reduce this whole thing to a single line. That basically means you provide a list of checks and as soon as one is true it returns that and stops evaluating. That would look something like this:

```
return this.trajectory == null || this.commandGroup == null || this.commandGroup.isFinished();
```

**Comment by paulelliott** [comment]
I think the code here looks really good, but I would love to see you write some JUnit tests for this. Take a crack at writing some basic tests for the `end` and `isFinished` functions. You can see a basic test example in #44: https://github.com/Creekside-Robotics/Creekside-8775-Robotics-2022-Offseason/pull/44/files#diff-77256b8d478a9ebd39c47ac98fc97e82028eb8e2c6bf6210c166ed06e88ba7be

**Comment by Catalyst4222** [comment]
> I think the code here looks really good, but I would love to see you write some JUnit tests for this. Take a crack at writing some basic tests for the `end` and `isFinished` functions. You can see a basic test example in #44: [#44 (files)](https://github.com/Creekside-Robotics/Creekside-8775-Robotics-2022-Offseason/pull/44/files#diff-77256b8d478a9ebd39c47ac98fc97e82028eb8e2c6bf6210c166ed06e88ba7be)

How would I test it? The wpilib docs don't have an example on how to do tests for commands, and I can't think of a way to test the command without needing access to the robot

**Comment by paulelliott** [comment]
We're starting to write unit tests in our project to exercise some of this code an in an automated way. The basic idea is you set up the object in certain ways and assert that the outcome is correct. I linked to a file in Joseph's PR where he wrote some tests for another class, which you can use as an example.

I doubt you'll find anything about this in the wpilib docs. This is a programming practice that I don't think is commonly found in the robotics community, although doing it will give you all a distinct advantage.

**Review by josephhadad05 — COMMENTED** [review]
Functionality wise, this code looks great. I think with the changes Mr. Elliot requested; you can move into unit tests either in this pr or a new one.

**Comment by Catalyst4222** [comment]
> We're starting to write unit tests in our project to exercise some of this code an in an automated way. The basic idea is you set up the object in certain ways and assert that the outcome is correct. I linked to a file in Joseph's PR where he wrote some tests for another class, which you can use as an example.

I looked at the test there, but the parts of the subsystem tested were things that could be done completely independently. I don't think there's any part of this command that could be tested without the required subsystems

> I doubt you'll find anything about this in the wpilib docs. This is a programming practice that I don't think is commonly found in the robotics community, although doing it will give you all a distinct advantage.

There's an example on how to test a subsystem, but unfortunately nothing on how to test commands

> Functionality wise, this code looks great. I think with the changes Mr. Elliot requested; you can move into unit tests either in this pr or a new one.

I'll add it in a separate pr, once I figure out a good way to test it

**Pull request merged**